README.md
# StringUtils

A set of common string functions, loosely based on LambdaMOO $string_utils. It uses [sevenecks/ansi](https://github.com/SevenEcks/ansi) to add color where applicable though you can pass your own colorizer in via dependency injection.

## Installation

Via Composer

```bash
composer require sevenecks/string-utils
```

## Usage

```php
require_once __DIR__ . '/../vendor/autoload.php';
use SevenEcks\StringUtils\StringUtils;
use SevenEcks\Ansi\Colorize;

$su = new StringUtils;
$test_string = 'This is a test';
// echo out test string
$su->tell($test_string);
$su->setLineLength(10);
// echo out a string, ignores line len
$su->tell($su->colorize->red("This is a test of a long red string!"));
$su->setSplitMidWord(true);
// echo using line length and color color
$su->tell($su->colorize->red($su->wordWrap("This is a test of a long blue word wrapped string which breaks mid word!")));
$su->setSplitMidWord(false);
// echo using length and color color
$su->tell($su->colorize->blue($su->wordWrap("This is a test of a long blue word wrapped string which breaks at a word!")));
// set the line length break string to br instead of default of \n
$su->setBreakString("<br />");
$su->tell($su->colorize->red($su->wordWrap("This is a test of a long red string!")));
$su->tell("Not colored.");
// mass colorize a string wrapped by line
$su->tell_lines($su->massColor($su->lineWrap("This is a test of a long line wrapped red strings!"), 'blue'));
// break at specific linelen
$su->tell_lines($su->lineWrap('123456789123456789123456789'));
// formatting of left, right, and center
$su->tell($su->left("TEST", 10) . $su->left("TEST1", 10));
$su->tell($su->right("TEST", 10) . $su->right("TEST1", 10));
$su->tell($su->center("TEST", 10) . $su->center("TEST1", 10));
// update line len
$su->setLineLength(180);
// center
$su->tell($su->center("CENTER NO ARGS"));
$su->tell($su->center("CENTER W/ ARGS", 30));
// alerts crits and warnings for debugging
$su->alert("This is an alert!");
$su->warning("This is a warning");
$su->critical("This is critical!");
// using tostr to combine args into a string
$su->tell($su->tostr($su->center("THIS EXAMPLE", 10), ' ', $su->center("USES TOSTR", 10), ' ', 1,2,3));
$su->tell($su->tostr('1',2,'3','asdf',['a', 'b' => 'c', 1]));
```
![Example Output](https://github.com/SevenEcks/string-utils/blob/master/images/example.png "Example Output")

## Change Log
Please see [Change Log](CHANGELOG.md) for more information.

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
## 0.1.1 - 2018-10-18

* Updated Readme
* Updated Changelog

## 0.1.0 - 2018-10-18

* Added massColor method to support colorizing an array
* modified line wrapping to return it's value instead of echoing
* Updated example.php to better illustrate uses

## 0.0.15 - 2018-04-04

- Moved Tableify to it's own package sevenecks/tableify
- Moved maxMulti... to it's own package (tableify) along with rowsize function for tablify

## 0.0.14 - 2018-04-03 5:45PM

- Added Tableify method to make a multidimensional array into a table
- Updated tostr to properly handle arrays using RecursiveIteratorIterator
- Updated example files with test data
- Added fill method which is a wrapper for str_pad with no starting string to create strings of a specific length
- Added maxMultidimensionalArrayStringSize method which slices through an array and gets the largest strlen for each index

## 0.0.13 - 2018-03-23 12:07PM

- Added optional param to ->center for length, defaulting to a portion of the line_len

## 0.0.12 - 2018-03-21 4:45PM

- Added optional param to ->tell to allow value to be returned instead of echod

## 0.0.11 - 2018-03-20 3:47PM

- Custom Colorize can be passed in through dependency injection
- Updated ReadMe
- Updated example/example.php
- Added new example2.php

## 0.0.10 - 2017-12-22 12:01PM

- Added SevenEcks/Ansi/Colorize to StringUtils class

## 0.0.9 - 2017-12-22 11:49AM

- Added warning, alert and critical strings

## 0.0.8 - 2017-12-12 11:51AM

- Updated README.md

## 0.0.7 - 2017-12-12 11:26AM

- Updated README.md

## 0.0.6 - 2017-12-12 11:10AM

- Updated sevenecks/ansi to ^1.0

## 0.0.5 - 2017-12-12 11:06AM

- Removed tests folder

## 0.0.4 - 2017-12-12 11:04AM

- Updated minimum stability to stable

## 0.0.3 - 2017-12-12 11:02AM

- Removed PHPUnit from composer.json as we aren't using it

## 0.0.2 - 2017-12-12 11:00AM

- Updated README
- Added example.png image
- Added example directory
- Moved test.php to example directory
- Renamed test.php to example.php

## 0.0.1 - 2017-12-12 10:51AM

- First commits


example/example.php
<?php
require_once __DIR__ . '/../vendor/autoload.php';
use SevenEcks\StringUtils\StringUtils;
use SevenEcks\Ansi\Colorize;

$su = new StringUtils;
$test_string = 'This is a test';
// echo out test string
$su->tell($test_string);
$su->setLineLength(10);
// echo out a string, ignores line len
$su->tell($su->colorize->red("This is a test of a long red string!"));
$su->setSplitMidWord(true);
// echo using line length and color color
$su->tell($su->colorize->red($su->wordWrap("This is a test of a long blue word wrapped string which breaks mid word!")));
$su->setSplitMidWord(false);
// echo using length and color color
$su->tell($su->colorize->blue($su->wordWrap("This is a test of a long blue word wrapped string which breaks at a word!")));
// set the line length break string to br instead of default of \n
$su->setBreakString("<br />");
$su->tell($su->colorize->red($su->wordWrap("This is a test of a long red string!")));
$su->tell("Not colored.");
// mass colorize a string wrapped by line
$su->tell_lines($su->massColor($su->lineWrap("This is a test of a long line wrapped red strings!"), 'blue'));
// break at specific linelen
$su->tell_lines($su->lineWrap('123456789123456789123456789'));
// formatting of left, right, and center
$su->tell($su->left("TEST", 10) . $su->left("TEST1", 10));
$su->tell($su->right("TEST", 10) . $su->right("TEST1", 10));
$su->tell($su->center("TEST", 10) . $su->center("TEST1", 10));
// update line len
$su->setLineLength(180);
// center
$su->tell($su->center("CENTER NO ARGS"));
$su->tell($su->center("CENTER W/ ARGS", 30));
// alerts crits and warnings for debugging
$su->alert("This is an alert!");
$su->warning("This is a warning");
$su->critical("This is critical!");
// using tostr to combine args into a string
$su->tell($su->tostr($su->center("THIS EXAMPLE", 10), ' ', $su->center("USES TOSTR", 10), ' ', 1,2,3));
$su->tell($su->tostr('1',2,'3','asdf',['a', 'b' => 'c', 1]));


src/StringUtils.php
<?php
namespace SevenEcks\StringUtils;

use SevenEcks\Ansi\Colorize;
use SevenEcks\Ansi\ColorInterface;

/**
 * This class is a utility package for manipulating strings in ways that. Specifically designed
 * for formatting terminal output and log output
 *
 * @author Brendan Butts <bbutts@stormcode.net>
 *
 */
class StringUtils
{
    protected $line_length = 180;
    protected $split_mid_word = false;
    protected $break_string = "\n";
    protected $eol_string = "\n";
    public $colorize = null;

    /**
     * Construct the StringUtils object, accepting a class that defines the
     * ColorizeInterface, if it is not passed in, the constructor injects one
     * on it's own.
     *
     * @param interface $colorize;
     * @return void
     */
    public function __construct(ColorInterface $colorize = null)
    {
        if (!$colorize) {
            $colorize = new Colorize;
        }
        $this->colorize = $colorize;
    }

    /**
     * Display a string on the left, and pad it on the right $length
     * characters using $filler.
     *
     * @param string $string
     * @param int $length
     * @param optional string $filler
     * @return string
     */
    public function left($string, $length, $filler = ' ') 
    {
        return str_pad($string, $length, $filler, STR_PAD_RIGHT); 
    }

    /**
     * Display a string on the right, and pad it on the left $length
     * characters using $filler.
     *
     * @param string $string
     * @param int $length
     * @param optional string $filler
     * @return string
     */
    public function right($string, $length, $filler = ' ') 
    {
        return str_pad($string, $length, $filler, STR_PAD_LEFT); 
    }

    /**
     * Display a string in the center (passed in, or determined by the 
     * line_length / 2 - the length of the string being presented / 2
     * and pad it on the left $length and right as close to equal
     * characters using $filler / 2.
     *
     * @param string $string
     * @param optional int $length
     * @param optional string $filler
     * @return string
     */
    public function center($string, $length = null, $filler = ' ')
    {
        if (is_null($length)) {
            $length = ($this->line_length / 2) - (strlen($string) / 2);
        }
        return str_pad($string, $length, $filler, STR_PAD_BOTH);
    }

    /**
     * Create a string of $length with the filler characters $filler.
     * This function will return a string of length $length even if
     * $filler is greater than that length.
     *
     * @param int $length
     * @param string $filler
     * @return string
     */
    public function fill(int $length, string $filler = ' ')
    {
        return str_pad('', $length, $filler);
    }

    /**
     * Break a string at $this->line_length
     * and return an array of strings
     *
     * @param string $string
     * @return array
     */
    public function lineWrap($string)
    {
        $lines = str_split($string, $this->line_length);
        return $lines;
    }

    /**
     * Echo or return a string ending in $this->eol_string
     *
     * @param string $string
     * @param bool $echo
     * @return string
     */
    public function tell(string $string, $echo = true)
    {
        $full_string = $string . $this->eol_string;
        if ($echo) {
            echo $full_string;
            return;
        }
        return $full_string;
    }

    /** 
     * Take n arguments and combine them into a single
     * string and return that string
     * @param string $strings
     * @param ... $strings
     * @return string
     */
    public function tostr(...$strings)
    {
        $new_string = '';
        foreach ($strings as $temp) {
            switch (gettype($temp)) {
                case "array":
                    $it = new \RecursiveIteratorIterator(new \RecursiveArrayIterator($temp));
                    foreach($it as $v) {
                        $temp_str .= ' ' . $this->tostr($v);
                    }
                    break;
                default:
                    $temp_str = (string)$temp;
                    break;
            }
            $new_string .= $temp_str;
        }
        return trim($new_string);
    }

    /**
     * Take an array of $lines and execute
     * $this->tell on each line
     *
     * @return none
     */
    public function tell_lines(array $lines)
    {
        foreach ($lines as $line) {
            $this->tell($line);
        }
    }

    /**
     * Return a string wrapped either at a word, or close to the words completion
     * using the break_string defined on the object
     *
     * @param string $string
     * @return none
     */
    public function wordWrap($string)
    {
        return wordwrap($string, $this->line_length, $this->break_string, $this->split_mid_word);
    }

    /**
     * Setter for $this->split_mid_word
     *
     * @param bool $split_mid_word
     * @return none
     */
    public function setSplitMidWord(bool $split_mid_word)
    {
        $this->split_mid_word = $split_mid_word;
    }

    /**
     * Getter for $this->get_split_word
     *
     * @return bool
     */
    public function getSplitMidWord()
    {
        return $this->split_mid_word;
    }

    /**
     * Setter for $this->line_length which defines
     * the line length for wrapping
     *
     * @param int $new_line_length
     * @return none
     */
    public function setLineLength(int $new_line_length)
    {
        $this->line_length = $new_line_length;
    }

    /**
     * Getter for $this->line_length
     *
     * @return int
     */
    public function getLineLength()
    {
        return $this->line_length;
    }

    /**
     * Setter for $this->break_string which defines what wordwrap
     * uses for line breaks when executing
     *
     * @param string $break_string
     * @return none
     */
    public function setBreakString(string $break_string)
    {
        $this->break_string = $break_string;
    }

    /**
     * Getter for $this->break_string
     *
     * @return string
     */
    public function getBreakString()
    {
        return $this->break_string;
    }

    /**
     * Getter for $this->eol_string which defines what tell()
     * uses when echoing out a string
     *
     * @return string
     */
    public function getEolString()
    {
        return $this->eol_string;
    }

    /**
     * Setter for $this->eol_string
     *
     * @return none
     */
    public function setEolString(string $eol_string)
    {
        $this->eol_string = $eol_string;
    }

    //TODO: do something better with this below code
    /**
     * Display an alert string
     *
     * @param string $string
     * @return none
     */
    public function alert(string $string)
    {
        $this->tell('[' . $this->colorize->cyan('ALERT') . '] ' .  $string);
    }

    /**
     * Display a warning string
     *
     * @param string $string
     * @return none
     */
    public function warning(string $string)
    {
        $this->tell('[' . $this->colorize->yellow('WARNING') . '] ' . $string);
    }

    /**
     * Display a critical string
     *
     * @param string $string
     * @return none
     */
    public function critical(string $string)
    {
        $this->tell('[' . $this->colorize->red('CRITICAL') . '] ' . $string);
    }

    /**
     * Mass colorize an array of strings
     *
     * @param array $strings
     * @param string $color
     * @return array
     */
    public function massColor(array $strings, string $color)
    {
        $len_array = count($strings);
        for ($i = 0; $i < $len_array; $i++) {
            $strings[$i] = $this->colorize->{$color}($strings[$i]);
        }
        return $strings;
    }
}


example/example2.php
<?php
require_once __DIR__ . '/../vendor/autoload.php';
use SevenEcks\StringUtils\StringUtils;
use SevenEcks\Ansi\Colorize;

$colorize = new Colorize;
$su = new StringUtils($colorize);
$test_string = 'This is a test';
$su->tell($test_string);
$su->setLineLength(10);
$su->tell(Colorize::red("This is a test of a long red string!"));
$su->setSplitMidWord(true);
$su->tell(Colorize::blue("This is a test of a long blue word wrapped string which breaks mid word!"));
$su->setSplitMidWord(false);
$su->tell(Colorize::blue("This is a test of a long blue word wrapped string which breaks at a word!"));
$su->setBreakString("<br />");
$su->tell(Colorize::red("This is a test of a long red string!"));
$su->tell("Not colored.");
$su->tell_lines($su->lineWrap("This is a test of a long red wrapped string!"));
$su->tell_lines($su->lineWrap('123456789123456789123456789'));
$su->tell($su->left("TEST", 10) . $su->left("TEST1", 10));
$su->tell($su->right("TEST", 10) . $su->right("TEST1", 10));
$su->tell($su->center("TEST", 10) . $su->center("TEST1", 10));
// using tostr to combine args into a string
$su->tell($su->tostr($su->center("THIS EXAMPLE", 10), ' ', $su->center("USES TOSTR", 10), ' ', 1,2,3));
$su->alert("This is an alert!");
$su->warning("This is a warning");
$su->critical("This is critical!");
// testing tableify and array to string conversion
$data = [
    ['Name', 'Date', 'Phone', 'Age'], 
    ['Altec Lansing', '03/22/18', '617-555-0584', '30'],
    ['Fack', '03/22/18', '508-555-0584', '24'],
    ['Seven Ecks', '03/22/18', '+1-888-555-0584', '100'],
    ['CK', '03/22/18', 'N/A', '33'],
    ['Jason Jasonson', '03/22/18', '978-555-0584', '34'],
    ['Waxillium Wick', '03/22/18', '978-555-0584', '34'],
    ['Ruby Reide', '03/22/18', '978-555-0584', '34'],
    ['Rex Gold', '03/22/18', '978-555-0584', '34'],
    ['Juicy Vee', '03/22/18', '978-555-0584', '34'],
];
$su->tell_lines($su->tableify($data, 'left', 1, '|', '-', '-'));
$su->tell_lines($su->tableify($data, 'center', 1, '|', '-', '-'));
$su->tell_lines($su->tableify($data, 'right', 1, '|', '-', '-'));
// testing tostr of an array
$su->tell("tostr of an array:");
$su->tell($su->tostr($data));


